Source/JavaScriptCore/runtime/ProxyObject.cpp

```diff
@@ -82,6 +86,64 @@ void ProxyObject::finishCreation(VM& vm, JSGlobalObject* globalObject, JSValue t
     internalField(Field::Handler).set(vm, this, handler);
 }
 
+JSObject* ProxyObject::getHandlerTrap(JSGlobalObject* globalObject, JSObject* handler, CallData& callData, const Identifier& ident, HandlerTrap trap)
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    auto ensureIsCallable = [&](JSValue value) -> JSObject* {
+        if (value.isUndefinedOrNull())
+            return nullptr;
+
+        callData = JSC::getCallData(value);
+        if (callData.type == CallData::Type::None) {
+            throwTypeError(globalObject, scope, makeString("'", String(ident.impl()), "' property of a Proxy's handler should be callable"));
+            return nullptr;
+        }
+
+        return asObject(value);
+    };
+
+    JSObject* objectPrototype = globalObject->objectPrototype();
+
+    bool isCacheValid = handler->structureID() == m_handlerStructureID.value() && objectPrototype->structureID() == m_handlerPrototypeStructureID.value();
+    if (isCacheValid) {
+        PropertyOffset offset = m_handlerTrapsOffsetsCache[static_cast<uint8_t>(trap)];
+        if (offset == invalidOffset)
+            return nullptr;
+        if (offset != emptyHandlerTrapCache)
+            return ensureIsCallable(handler->getDirect(offset));
+    } else
+        clearHandlerTrapsOffsetsCache();
```

Comment:
We should do this only when `m_handlerStructureID` is set.

Reply:
Done!